sdk/tests/conformance2/reading/read-pixels-upon-pixel-pack-buffer.html

```diff
@@ -0,0 +1,219 @@
+void main()
+{
+    color = colorIn;
+    gl_Position = vec4(pos.xyz, 1.0);
+}
+</script>
+
+<script id="fshader" type="x-shader/x-fragment">
+precision mediump float;
+varying vec4 color;
+
+void main()
+{
+    gl_FragColor = color;
+}
+</script>
+</head>
+<body>
+<canvas id="example" width="32" height="32"></canvas>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+"use strict";
+
+function validatePixelPackBufferAndParameters()
+{
+    debug("Validate PIXEL_PACK buffer and readPixels' parameters");
+    gl.clearColor(0, 0, 0, 1);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+    gl.pixelStorei(gl.PACK_ALIGNMENT, 1);
+
+    var size = width * height * 4;
+    var buffer = gl.createBuffer();
+    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, buffer);
+    gl.bufferData(gl.PIXEL_PACK_BUFFER, size, gl.STATIC_DRAW);
+    var array = new Uint8Array(size);
+
+    debug("PIXEL_PACK buffer is bound");
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, array);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "pixel pack buffer should not be bound");
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    debug("Validate the offset of PIXEL_PACK buffer and buffer size");
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, -1);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE, "offset < 0");
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, size);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "offset > buffer size");
+    gl.readPixels(0, 0, 2, 1, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "pixel pack buffer is not large enough");
+
+    debug("Validate the reading area of framebuffer");
+    gl.readPixels(-1, -2, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "reading pixels outside of the framebuffer should succeed.");
+    gl.readPixels(2, 1, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "reading pixels outside of the framebuffer should succeed.");
+    gl.readPixels(2, 1, -1, -1, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE,
+                        "reading pixels with negative width / height should generate INVALID_VALUE.");
+
+    debug("no PIXEL_PACK buffer bound");
+    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, null);
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, array);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "no pixel pack buffer bound");
+
+    gl.deleteBuffer(buffer);
+}
+
+function calculatePaddingBytes(bytesPerPixel, packAlignment, width)
+{
+    var padding = 0;
+    switch (packAlignment) {
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+        padding = (bytesPerPixel * width) % packAlignment;
+        if (padding > 0)
+            padding = packAlignment - padding;
+        break;
```

Comment:
Why not just return "padding" here, rather than breaking out and returning it at the end of the function?